### PR TITLE
WS-1738: Course Recommendation endpoint QA

### PR DESCRIPTION
**Pull request by Vyraal1**
Refined to use querysets for de-duping and exclusion of calling course. Added an additional test to check for matching both subjects and organizations.

### CODE REVIEW

File: course_discovery/apps/course_metadata/models.py [hunk 1]
```diff
@@ -1134,28 +1134,16 @@ def recommendations(self):
...
-                seen.add(course)
-
-        return deduped_courses
+        in priority of A over B
+        """
+        recommended_courses = (
+            Course.objects.filter(
+                subjects__in=self.subjects.all(),
+                authoring_organizations__in=self.authoring_organizations.all()
+            ) | Course.objects.filter(
```
Comment (rgraber) [inline course_discovery/apps/course_metadata/models.py]: does the '|' mean there's an OR in the query, or we'll only look for matching programs if we don't find any of the same subj/org?

Comment (rgraber) [reply course_discovery/apps/course_metadata/models.py]: also were you able to test this with an empty subject list? that's an edge case so NBD if not, but I'm curious.

Comment (Vyraal1) [reply course_discovery/apps/course_metadata/models.py]: It's a SQL OR, so it'll add a course as long as it matches either: 
`prog1 OR prog2....`
OR
`(subject1 OR subject2...) AND (org1 OR org2...)`

**Review by rgraber — COMMENTED** [review]
We need to return the ones from the same program first in the list

**Comment by Vyraal1** [comment]
Should be ready now, had to use python native list since `union` and using other methods of maintain order such as `annotate` with `order_by` did not preserve the order.

File: course_discovery/apps/course_metadata/models.py [hunk 2]
```diff
@@ -1134,28 +1134,23 @@ def recommendations(self):
...
+            .distinct()
+            .all())
+
+        subject_org_courses = list(
+            Course.objects.filter(
+                subjects__in=self.subjects.all(),
+                authoring_organizations__in=self.authoring_organizations.all()
+            ).exclude(key=self.key)
+            .distinct()
+            .all())
```
Comment (waheedahmed) [inline course_discovery/apps/course_metadata/models.py:1151]: Can we combine these two queries using `Q`?

**Review by waheedahmed — APPROVED** [review]
Just one question otherwise looks good 👍

File: course_discovery/apps/course_metadata/models.py (hunk 2 above)
Comment (Vyraal1) [reply course_discovery/apps/course_metadata/models.py:1151]: I tried originally using Q's, but I had trouble preserving the order for `program_courses` over `subject_org_courses`. 

Approaches I tried were using `annotate` to have a `rec_order` with a value of 1 for the program courses, and a value of 0 for the subject courses and ordering by that, but I couldn't get it and eventually decided the trade-off wasn't worth spending more time on it for a short-term experiment. 

If the experiment is a win, there will be more emphasis on further refining the query.

**Review by waheedahmed — APPROVED** [review]
👍 once build is green

**Review by albemarle — APPROVED** [review]

**Review by srwang — APPROVED** [review]
